src/tools/cephfs_mirror/PeerReplayer.cc

```diff
@@ -1319,6 +1321,164 @@ int PeerReplayer::do_synchronize(const std::string &dir_root, const Snapshot &cu
+    }
+
+    dout(20) << ": " << sync_queue.size() << " entries in queue" << dendl;
+    const auto &queue_entry = sync_queue.front();
+    epath = queue_entry.path;
+    dout(20) << ": syncing entry, path=" << epath << dendl;
+    r = ceph_open_snapdiff(m_local_mount, dir_root.c_str(), epath.c_str(),
+                           stringify((*prev).first).c_str(), current.first.c_str(), &sd_info);
+    if (r != 0) {
+      derr << ": failed to open snapdiff, r=" << r << dendl;
+      return r;
+    }
+    while (0 < (r = ceph_readdir_snapdiff(&sd_info, &sd_entry))) {
+      if (r < 0) {
+        derr << ": failed to read directory=" << epath << dendl;
+        ceph_close_snapdiff(&sd_info);
+        return r;
+      }
+
+      //New entry found
+      nname = sd_entry.dir_entry.d_name;
+      if ("." == nname || ".." == nname)
+        continue;
+      // create path for the newly found entry
+      npath = entry_diff_path(epath, nname);
+      nabs_path = entry_diff_path(dir_root, npath);
+
+      struct ceph_statx cstx;
+      r = ceph_statx(sd_info.cmount, nabs_path.c_str(), &cstx,
+                     CEPH_STATX_MODE | CEPH_STATX_UID | CEPH_STATX_GID |
+                     CEPH_STATX_SIZE | CEPH_STATX_ATIME | CEPH_STATX_MTIME,
+                     AT_STATX_DONT_SYNC | AT_SYMLINK_NOFOLLOW);
+      if (r < 0) {
+        // can't stat, so it's a deleted entry.
+        if (DT_DIR == sd_entry.dir_entry.d_type) { // is a directory
+          r = cleanup_remote_dir(dir_root, npath, fh);
+          if (r < 0) {
+            derr << ": failed to remove directory=" << nabs_path << dendl;
+            break;
+          }
+        }
+        else { // is a file
+          r = ceph_unlinkat(m_remote_mount, fh.r_fd_dir_root, npath.c_str(), 0);
+          if (r < 0) {
+            break;
+          }
+        }
+      } else {
+        // stat success, update the existing entry
+        struct ceph_statx tstx;
+        int rstat_r = ceph_statx(m_remote_mount, nabs_path.c_str(), &tstx,
+                                 CEPH_STATX_MODE | CEPH_STATX_UID | CEPH_STATX_GID |
+                                 CEPH_STATX_SIZE | CEPH_STATX_ATIME | CEPH_STATX_MTIME,
+                                 AT_STATX_DONT_SYNC | AT_SYMLINK_NOFOLLOW);
+        if (S_ISDIR(cstx.stx_mode)) { // is a directory
+          //cleanup if it's a file in the remotefs
+          if (!rstat_r && !S_ISDIR(tstx.stx_mode)) {
+            r = ceph_unlinkat(m_remote_mount, fh.r_fd_dir_root, npath.c_str(), 0);
+            if (r < 0) {
+              derr << ": Error in directory sync. Failed to remove file="
+                   << nabs_path << dendl;
+              break;
+            }
+          }
+          r = remote_mkdir(npath, cstx, fh);
+          if (r < 0) {
+            break;
+          }
+          // push it to sync_queue for later processing
+          sync_queue.emplace(SyncDiffEntry(sd_entry, npath));
+        } else { // is a file
+          bool need_data_sync = true;
```

Comment:
Under this case, doesn't there need to be handling if the remote entry is not a file or is that handled elsewhere?

Reply:
It should be at https://github.com/ceph/ceph/pull/54633/files#diff-76c97c0667b3b381c954391b23d3b0845c63d026f593dbc93663e9d4a217f38aR1463.